opentracing/src/main/java/opentracing/SpanPropagator.java

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright 2016 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package opentracing;
+
+import java.util.Map;
+
+/**
+ * Encodes or Decodes a {@linkplain Span} in binary or text formats.
+ *
+ * <p>The toXXX methods are expected to serialize spans into a pair of values representing
+ * separately the span / span identity, and the trace attributes. This is done specifically
+ * for binary protocols that may represent tracing identity in a dedicated fixed-length slot in the
+ * binary message format, so that it can be inspected efficiently by the middleware / routing layers
+ * without parsing the whole message.
+ */
+public interface SpanPropagator {
+
+  /**
+   * Encodes the span into a binary representation of the span's identity and trace
+   * attributes.
+   */
+  byte[] toBinary(Span span);
+
+  /**
+   * Decodes a span from a binary representation of the span.
+   *
+   * @throws IllegalArgumentException if the encoded data is malformed.
+   */
+  Span fromBinary(byte[] encoded);
```

Comment:
@michaelsembwever I think we have this as unresolved issue https://github.com/opentracing/opentracing.github.io/issues/22.  Basically, there are layers that may want to add stuff to a span, but not affect the span's timestamp/duration, which is implied by the `finish` method. But at this point I think we indeed expect `fromBinary` to always return a started span that must be finished (as that's the only way to cause it to be reported).

But there's a second problem here. The former `join` method was abstracting away the difference in how tracing systems model RPCs, Dapper with two spans, Zipkin with one span. The tracer was deciding whether `joinFromBinary` would return the same spanID (Zipkin) or a child span ID (Dapper), without forcing the end user to make that choice. Expecting the same semantics from `fromBinary